### PR TITLE
Forgot password and reset password

### PR DESCRIPTION
**Pull request by abhishek77in**

### CODE REVIEW

File: app/views/users/forgot_password.haml [hunk 1]
```diff
@@ -0,0 +1,13 @@
+%h2
+  Reset password Page
+= form_tag '/users/reset_password' do
```
Comment (abhishek77in) [inline app/views/users/forgot_password.haml]: Use rails route helpers

File: app/models/user.rb [hunk 2]
```diff
@@ -16,4 +17,24 @@ class User < ApplicationRecord
...
+  def password_generator
+    [*('a'..'z')].sample(4).join + rand(1000..9999).to_s
+  end
+
+  def change_password
+    new_password=password_generator
+    puts new_password
+    self.password=new_password
+    self.save
+    puts"New pass aagya"
```
Comment (abhishek77in) [inline app/models/user.rb]: Avoid checkin code used for debugging and development.

File: app/models/user.rb [hunk 3]
```diff
@@ -16,4 +17,24 @@ class User < ApplicationRecord
...
+    self.send_reset_password_sms
```
Comment (abhishek77in) [inline app/models/user.rb]: `change_password` method on `User` should only change user password.

File: app/models/user.rb [hunk 4]
```diff
@@ -16,4 +17,24 @@ class User < ApplicationRecord
   def self.users_for_usg_reports?
     User.active.for_usg_reports.count.positive?
   end
+
+  def send_reset_password_sms
+    return if mobile_number.blank?
```
Comment (abhishek77in) [inline app/models/user.rb]: Check for mobile number in controller.

File: app/models/user.rb (hunk 3 above)
Comment (bajajerk) [reply app/models/user.rb]: If i make a method to send sms , and write after change_password send sms, then I wouldnt able to send new generated password to that method.

File: app/models/user.rb [hunk 5]
```diff
@@ -16,4 +17,22 @@ class User < ApplicationRecord
...
   end
+
+  def send_reset_password_sms(new_password)
+    return if mobile_number.blank?
+    sms = self.sms_messages.new(mobile_number: self.mobile_number, message: 'message')
+    sms.save
+    sms.deliver_later
+  end
+
+  def password_generator
```
Comment (abhishek77in) [inline app/models/user.rb]: Call this `random_password`

Comment (abhishek77in) [reply app/models/user.rb]: Make it a self method. `User.random_password`

File: app/models/user.rb [hunk 6]
```diff
@@ -16,4 +17,22 @@ class User < ApplicationRecord
...
+    [*('a'..'z')].sample(4).join + rand(1000..9999).to_s
+  end
+
+  def change_password
+    new_password=password_generator
+    self.password = new_password
+    self.save
+    self.send_reset_password_sms(new_password)
```
Comment (abhishek77in) [inline app/models/user.rb]: Move this to controller.

File: app/models/user.rb [hunk 7]
```diff
@@ -16,4 +17,15 @@ class User < ApplicationRecord
   def self.users_for_usg_reports?
     User.active.for_usg_reports.count.positive?
   end
+
+  def send_reset_password_sms(new_password)
```
Comment (abhishek77in) [inline app/models/user.rb]: Rename the method to `send_new_password`.

File: app/models/user.rb [hunk 8]
```diff
@@ -16,4 +17,15 @@ class User < ApplicationRecord
...
+    return if mobile_number.blank?
```
Comment (abhishek77in) [inline app/models/user.rb]: Do not check for mobile number here, already checked in controller.

File: app/controllers/users_controller.rb [hunk 9]
```diff
@@ -20,6 +21,23 @@ def create
...
+  end
+
+  def reset_password
+    user = User.find_by_username(params[:username])
+    return redirect_to '/', notice: 'Phone Number not added.' if user.mobile_number.blank?
+    new_password = random_password
+    if(user.send_reset_password_sms(new_password))
+      user.change_password(new_password)
+    end
+    return redirect_to '/'
```
Comment (abhishek77in) [inline app/controllers/users_controller.rb]: Use `root_path` instead of hard coding the routes.

Comment (abhishek77in) [reply app/controllers/users_controller.rb]: `return` not required.

File: app/controllers/users_controller.rb [hunk 10]
```diff
@@ -20,6 +21,23 @@ def create
...
+  def forgot_password
+  end
+
+  def random_password
+    [*('a'..'z')].sample(4).join + rand(1000..9999).to_s
+  end
+
+  def reset_password
+    user = User.find_by_username(params[:username])
+    return redirect_to '/', notice: 'Phone Number not added.' if user.mobile_number.blank?
```
Comment (abhishek77in) [inline app/controllers/users_controller.rb]: Use `alert` instead of `notice`.

Comment (abhishek77in) [reply app/controllers/users_controller.rb]: check if `return` is absolutely required here.

**Comment by abhishek77in** [comment]
@bajajerk Kindly refer to code climate - https://codeclimate.com/github/abhishek77in/azurite/pull/68

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -20,6 +21,23 @@ def create
     end
   end
 
+  def forgot_password
+  end
+
+  def random_password
```
Comment (abhishek77in) [inline app/controllers/users_controller.rb]: Move this under `private` controller methods.

File: app/models/user.rb (hunk 7 above)
Comment (bajajerk) [reply app/models/user.rb]: Changed

File: app/models/user.rb (hunk 8 above)
Comment (bajajerk) [reply app/models/user.rb]: Done

File: app/controllers/users_controller.rb (hunk 9 above)
Comment (bajajerk) [reply app/controllers/users_controller.rb]: Done

File: app/controllers/users_controller.rb (hunk 10 above)
Comment (bajajerk) [reply app/controllers/users_controller.rb]: Done

File: app/controllers/users_controller.rb (hunk 11 above)
Comment (bajajerk) [reply app/controllers/users_controller.rb]: Done

**Pull request merged**